implement an example of lazy-loading (#3)

thisgeek opened the pull request:
Resolves (#3).

As implemented, the lazy-loading employed here leverages webpack 2's support
for the pending dynamic import syntax. Eslint and babel now include plugins and
configurations that accept the new syntax.

Here's how it works: The dynamic import requests the files needed for the
application's routes. It provides the promise returned by the templateProvider
function for each route. For the controller, the approach is more indirect. At
the time of configuration, the route knows its controller by name, but that is
all. Within a resolve function, the ocLazyLoad service injects the module that
registers the controller. Now, when the router looks up the controller by name
it finds it in the registry.

Note: Grouping the related imports in a common chunk requires the use of a
"magic comment", which in turn required a webpack upgrade.

thescientist13 commented on .babelrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "plugins": ["syntax-dynamic-import"]
```
fyi about #14

thescientist13 commented on src/views/home/home.config.js:
```diff
@@ -1,16 +1,20 @@
     url: '/home',
-    template: HomeTemplate,
-    controller: HomeController,
+    templateProvider: () => import(/* webpackChunkName: "home.route" */ './home.template.html'),
```
:mind-blown:

thisgeek replied:
Hehe. That was a fun one to write. I didn't know until this that ui-router's templateProvider could handle a promise. It's not clear to me from the [documentation](https://ui-router.github.io/ng1/docs/0.4.2/#/api/ui.router.state.$stateProvider#methods_state) that it would, but I saw some [examples in the wild](https://github.com/preciousDev/angularMaterial-webpack2-ocLazyLoad/blob/master/app-content/main/main.js#L16-L23) that showed me it could.

thescientist13 commented:
So couple thoughts @thisgeek 
1. What are your thoughts of the build creeping into the source code like this?  A necessary evil perhaps, but curious on your thoughts overall?  (like "injecting" `NODE_ENV`, using `require`, etc)
1. Can we leave one example without lazy-loading?  Might not be practical, but more useful as this is reference app in reality
1. If we leave this in, maybe speak to it in the README or a Developer Notes page in the repo's wiki, we can link from README under the development section?

thisgeek commented:
On 1, now that you mention it does seem "wrong" to have the source leave clues for the builder all over the place. If we rely on comment-formatted annotations to decide how to package the code, then:

* not all comments are equally removable, and
* predicting build results will based on data scattered throughout the codebase.

I wasn't prepared to mitigate those consequences. I might be able to make the second one manageable with a script, but if I'm going to go that far I'd rather make a plugin. It might get complicated. The plugin may need to identify routes and find the template and controller paired to it. Sounds like a breeding ground for edge cases. I search for a plugin that does that. Maybe we'll get lucky.

It also worries me that the documentation didn't say that [chunk names][] are only temporary. It sure _looks_ like a temporary solution.

[chunk names]: https://webpack.js.org/guides/code-splitting-async/#chunk-names

On 2, yeah. That's a good idea. I can see reason to not lazy-load the homepage.

On 3, that is also a good idea. I'll look into it.

thisgeek commented:
Another thought: the idea of a webpack plugin that can "read" ui-router configurations gets even heavier when I remembered that you can configure the providers to load templates and controllers however you might like (e.g. assuming a naming convention based on the name of the state).

thescientist13 commented:
Saw this was posted last night.  Thoughts?  Did you see any impact to the Webpack build times?
https://github.com/webpack/webpack/issues/4636
https://github.com/webpack/webpack/issues/4636#issuecomment-301920118

thisgeek commented:
@thescientist13 I don't see a meaningful difference in our case.

With `time yarn run build` on master:

```
real	0m57.035s
user	0m44.449s
sys	0m2.048s
```

The same command on feature/issue-3-oclazyload:

```
real	0m42.862s
user	0m43.575s
sys	0m1.241s
```

thisgeek commented:
@thescientist13 Granted, the issue clearly says: "Create a large number static imports in your project" and "Convert those imports into dynamic imports". To attempt to reproduce, we would have to define "large", and then make the appropriate mass of files.

thisgeek commented:
@thescientist13 Now I'm second-guessing. Should I have measured the speed of hot module reloading? I should probably have tried something close to what [ingro did](https://github.com/webpack/webpack/issues/4636#issuecomment-301830807).

thisgeek commented:
I ran a new test. I measured the times for hot module reloading. For both the master branch and the feature/issue-3-oclazyload branch, I:

1. started `yarn run develop`,
2. changed something arbitrary in `src/views/users/users.config.js`
3. changed it back
4. repeated steps 2 and 3
5. captured the output from webpack

Again, I didn't see any meaningful difference, but that may prove little. Keep in mind that at the time of this experiment the feature branch had only one file using `import()`.

I published the outputs. You can review both the results on [master](https://gist.github.com/thisgeek/931830c8d01f12c5b024a1b55dd593cd) and on the [feature branch](https://gist.github.com/thisgeek/678e3fd0af1ca87b2ac777b3764519b6).

thisgeek commented:
@thescientist13 As to the second comment, it looks like the performance loss is the price of doing business. Agreed?

thisgeek commented:
Updates:

* After some cursory searching, I did _not_ luck upon a plugin that groups imports related by a ui-router configuration. I could try spending a little more time on it, but if so I recommend opening a separate ticket.
* The homepage configuration now loads the controller and template normally.
* I haven't started the wiki documentation yet. That's next on my list.

thisgeek commented:
@thescientist13 I posted a draft to the [wiki][]:

[wiki]: https://github.com/kenzanmedia/angularjs-webpack-seed/wiki

I didn't know whether it would give you a chance to review when I hit "Save". It looks like it didn't. We can keep this thread open for your feedback.

thisgeek commented:
One last thing to add: the link from the README to the wiki. I'll post that shortly and you can consider the whole thing ready for another full review.

thescientist13 commented:
Looks good @thisgeek !
